sources/web/datalab/polymer/components/datalab-files/datalab-files.ts

```diff
@@ -258,7 +256,154 @@ class FilesElement extends Polymer.Element {
+    return this._createNewItem('directory');
+  }
+
+  /**
+   * The Jupyter contents API does not provide a way to create a new item with a specific
+   * name, it only creates new untitled files or directories in provided path (see
+   * https://github.com/jupyter/notebook/blob/master/notebook/services/contents/manager.py#L311).
+   * In order to offer a better experience, we create an untitled item in the current path,
+   * then rename it to whatever the user specified.
+   * 
+   * This method creates an input modal to get the user input, then calls the ApiManager to
+   * create a new notebook/directory at the current path, and fetches the updated list
+   * of files to redraw.
+   */
+  _createNewItem(type: string) {
+
+    // First, open a dialog to let the user specify a name for the notebook.
+    const inputOptions: DialogOptions = {
+      title: 'New ' + type, 
+      withInput: true,
+      inputLabel: 'Name',
+      okLabel: 'Create',
+    };
+
+    return Utils.getUserInputAsync(inputOptions)
+      .then((closeResult: DialogCloseResult) => {
+        // Only if the dialog has been confirmed with some user input, rename the
+        // newly created file. Then if that is successful, reload the file list
+        if (closeResult.confirmed && closeResult.userInput) {
+          let newName = this.currentPath + '/' + closeResult.userInput;
+          // Make sure the name ends with .ipynb for notebooks for convenience
+          if (type === 'notebook' && !newName.endsWith('.ipynb')) {
+            newName += '.ipynb';
+          }
+          return ApiManager.createNewItem(type, newName)
+            .then(() => this._fetchFileList());
+        } else {
+          return Promise.resolve(null);
+        }
+      }); // TODO: Handle create errors properly by showing some message to the user
+  }
+
+  /**
+   * Creates an input modal to get the user input, then calls the ApiManager to
+   * rename the currently selected item. This only works if exactly one item is
+   * selected.
+   */
+  _renameSelectedItem() {
+
+    const selectedIndices = this.$.files.getSelectedIndices();
+    if (selectedIndices.length === 1) {
+      const i = selectedIndices[0];
+      const selectedObject = this._fileList[i];
+
+      // Open a dialog to let the user specify the new name for the selected item.
+      const inputOptions: DialogOptions = {
+        title: 'Rename ' + selectedObject.type.toString(), 
+        withInput: true,
+        inputLabel: 'New name',
+        inputValue: selectedObject.name,
+        okLabel: 'Rename',
+      };
+
+      // Only if the dialog has been confirmed with some user input, rename the
+      // selected item. Then if that is successful, and reload the file list.
+      return Utils.getUserInputAsync(inputOptions)
+        .then((closeResult: DialogCloseResult) => {
+          if (closeResult.confirmed && closeResult.userInput) {
+            let newName = this.currentPath + '/' + closeResult.userInput;
```

Comment:
Should this include the same check for .ipynb extension as in the create method on line 295? If not, then newName can be const.